### ISSUE TITLE
[bug] HTML5 document encoding differs from HTML4

### ISSUE THREAD

**Issue by flavorjones** [sparklemotion/nokogiri#2801]
**Please describe the bug**

The encoding of an HTML5 document differs from the encoding of an HTML4 document:

```
Nokogiri::HTML4::Document.parse(File.open(SHIFT_JIS_HTML)).encoding == "Shift_JIS"
Nokogiri::HTML5::Document.parse(File.open(SHIFT_JIS_HTML)).encoding == "UTF-8"
```

I haven't had time to dig into why this is (and whether it's intended behavior), so I'm opening this issue to look into it later. cc @stevecheckoway 

**Help us reproduce what you're seeing**

```ruby
#! /usr/bin/env ruby

require 'nokogiri'
require_relative 'test/helper'

class Test < Nokogiri::TestCase
  describe "document encoding" do
    it "should be shift_jis" do
      # passes
      assert_equal("Shift_JIS", Nokogiri::HTML4::Document.parse(File.open(SHIFT_JIS_HTML)).encoding)

      # fails
      assert_equal("Shift_JIS", Nokogiri::HTML5::Document.parse(File.open(SHIFT_JIS_HTML)).encoding)
    end
  end
end
```

**Expected behavior**

I think these should both be the same?

**Comment by flavorjones** [comment]
There are a couple of things worth digging into here:

1. why isn't the HTML4 EncodingReader discovering the encoding in the "given a string" case?
2. looks like passing a File and an encoding isn't supported by HTML5 (it raises an exception)
3. can we make the HTML5 encoding detection match the HTML5 behavior? I think we might be able to use the EncodingReader if it's helpful.

**Comment by stevecheckoway** [comment]
Since Gumbo doesn't support anything other than UTF-8, it performs the standard [encoding detection pre-scan](https://html.spec.whatwg.org/multipage/parsing.html#determining-the-character-encoding) that browsers are supposed to perform to decide on the encoding and then uses that to convert to UTF-8 to pass to Gumbo.

I'm wasn't sure what the `encoding` property is supposed to return so I [set it to the encoding of the strings in the Document](https://github.com/sparklemotion/nokogiri/blob/4ba43be4dcf5fb0c5c56110360c47db06908c1a0/lib/nokogiri/html5/document.rb#L124) itself rather than the encoding of the input document. If we want the latter, that is probably easy to change.

I'm not sure how this is supposed to interact with `#to_html` or other serialization methods though. Should those produce strings in the Document's encoding? I know Ruby has a concept of internal and external encoding for streams but I don't know how those should interact with this.

Basically, I don't know what the correct behavior around encodings is supposed to be but since Gumbo only supports UTF-8, I punted on it.
